### PR TITLE
Utilize Primary key information, to support additional queries

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Closes [#6190](https://github.com/apache/arrow-datafusion/issues/6190).

# Rationale for this change

Primary key information specifies that some of the columns in the table, would consist of unique values. This contract enables us to support additional queries, currently cannot be run by datafusiuon.
Consider the query below
```sql
SELECT sn, amount
            FROM sales_global
            GROUP BY sales_global.sn
```
When `sn` is `PRIMARY KEY` (each `sn` will have unique values) we know that, all the columns the table `sales_global`  can be emitted after aggregation (since for each group all rows would have same value). In terms of functionality, above query and following query 
```sql
SELECT sn, amount
            FROM sales_global
            GROUP BY sn, amount
```
are same. The reason is that, since column `sn` already consists of unique values; both `GROUP BY sn, amount` and `GROUP BY sn` will produce groups that consist of single row. 
```sql
SELECT sn, amount
            FROM sales_global
            GROUP BY sales_global.sn
```
Above query can run in Postgre. However, datafusion can only emit `sn` after aggregation from the original table. With this PR above query also supported by Datafusion.

As part of this PR, we keep track of identifier key (where primary key is a special case) to accomplish this.  To illusturate what an identifier key is 
Consider table below
| a  | b |
| ---- | --- |
| 1  | USD    |
| 2 | USD    |
| 3    | TRY    |

In table above, `column a` consists of unique values, it is primary key. We store this information as 
```rust
IdentifierKeyGroup{
  identifier_key_indices: vec![0], 
  is_unique: true, 
  associated_indices: vec![0, 1])
}
```
this means that, when we know the value of `0th` column, we will know the values of `0th and 1st` columns deterministically. And each row will be unique.
As an another example, consider table below,
| a | b | 
| ----- | ----- | 
| 1  | USD |
| 1  | USD |
| 2 | USD |
| 2 | USD |
| 3    | TRY |
| 3    | TRY |

In the example above,  `column a` is not primary key, however, still knowing value of `a` enables us to know value of `b` deterministically. We encode this information as 
```rust
IdentifierKeyGroup{
  identifier_key_indices: vec![0], 
  is_unique: false, 
  associated_indices: vec![0, 1])
}
```
All of these information are propagated from the primary key information at the source. These analysis enables us to support complex queries. As an example, while we can support query below
```sql
SELECT r.sn, SUM(l.amount), r.amount
  FROM sales_global_with_pk AS l
  JOIN sales_global_with_pk AS r
  ON l.sn >= r.sn
  GROUP BY r.sn
  ORDER BY r.sn
```
we reject, following query
```sql
SELECT l.sn, r.sn, SUM(l.amount), r.amount
  FROM sales_global_with_pk AS l
  JOIN sales_global_with_pk AS r
  ON l.sn >= r.sn
  GROUP BY l.sn
  ORDER BY l.sn
```
because second query, uses an unassociated column(`r.sn`) with `l.sn` after group by, whereas, first query uses only associated columns(`r.sn`, `r.amount`) with `r.sn` during projection.

# What changes are included in this PR?

# Are these changes tested?
Yes, new tests that show supported functionality are added to the `groupby.slt` file.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_planner.rs [hunk 1]
```diff
@@ -1025,10 +1026,18 @@ impl DefaultPhysicalPlanner {
...
+                            let right_primary_keys = add_offset_to_primary_key(&right_primary_keys, left_field_indices.len());
+
+                            let primary_keys = left_primary_keys.into_iter().chain(right_primary_keys).collect();
 
                             // Construct intermediate schemas used for filtering data and
                             // convert logical expression to physical according to filter schema
-                            let filter_df_schema = DFSchema::new_with_metadata(filter_df_fields, HashMap::new())?;
+                            let filter_df_schema = DFSchema::new_with_metadata(filter_df_fields, HashMap::new())?.with_primary_keys(primary_keys);
                             let filter_schema = Schema::new_with_metadata(filter_fields, HashMap::new());
                             let filter_expr = create_physical_expr(
```
Comment (metesynnada) [inline datafusion/core/src/physical_planner.rs:1042]: I am not sure how this affects the filter expr. How can use dfschema primary key information without changing `create_physical_expr`?

**Comment by mustafasrepo** [comment]
Merged upstream

File: datafusion/core/src/physical_planner.rs [hunk 2]
```diff
@@ -1025,10 +1026,18 @@ impl DefaultPhysicalPlanner {
...
+
+                            // after join, indices from the right table will increase with the size of the left table. hence add offset
+                            let right_primary_keys = add_offset_to_primary_key(&right_primary_keys, left_field_indices.len());
+
+                            let primary_keys = left_primary_keys.into_iter().chain(right_primary_keys).collect();
 
                             // Construct intermediate schemas used for filtering data and
                             // convert logical expression to physical according to filter schema
-                            let filter_df_schema = DFSchema::new_with_metadata(filter_df_fields, HashMap::new())?;
+                            let filter_df_schema = DFSchema::new_with_metadata(filter_df_fields, HashMap::new())?.with_primary_keys(primary_keys);
```
Comment (metesynnada) [inline datafusion/core/src/physical_planner.rs:1040]:
```suggestion
                            let filter_df_schema = DFSchema::new_with_metadata(filter_df_fields, HashMap::new())?;
```

File: datafusion/expr/src/logical_plan/builder.rs [hunk 3]
```diff
@@ -1214,6 +1271,84 @@ pub fn union(left_plan: LogicalPlan, right_plan: LogicalPlan) -> Result<LogicalP
...
+    entries: &[usize],
+    proj_indices: &[usize],
+) -> Vec<usize> {
+    entries
+        .iter()
+        .filter_map(|val| proj_indices.iter().position(|proj_idx| proj_idx == val))
+        .collect()
+}
+
+/// Update identifier key indices, with index of the number in `field_indices`
```
Comment (mustafasrepo) [inline datafusion/expr/src/logical_plan/builder.rs:1285]:
```suggestion
/// Update identifier key indices, with index of the number in `proj_indices`
```

File: datafusion/expr/src/logical_plan/builder.rs [hunk 4]
```diff
@@ -1214,6 +1271,84 @@ pub fn union(left_plan: LogicalPlan, right_plan: LogicalPlan) -> Result<LogicalP
...
+/// If `proj_indices` is \[2, 5, 8\], and identifier key groups is \[5\] -> \[5, 8\] in the `df_schema`.
+/// return value will be \[1\] -> \[1, 2\]. This means that 1st index of the `field_indices`(5) is identifier key,
```
Comment (mustafasrepo) [inline datafusion/expr/src/logical_plan/builder.rs:1287]:
```suggestion
/// return value will be \[1\] -> \[1, 2\]. This means that 1st index of the `proj_indices`(5) is identifier key,
```

**Review by ozankabak — APPROVED** [review]
LGTM, let's write a PR body, choose an appropriate title and send upstream.

**Comment by mustafasrepo** [comment]
Upstream [PR](https://github.com/apache/arrow-datafusion/pull/7040)

**Comment by mustafasrepo** [comment]
Merged upstream

**Pull request closed without merging**